pipelex/tools/misc/path_utils.py

```
@@ -111,3 +111,16 @@ def clarify_path_or_url(path_or_uri: str) -> tuple[str | None, str | None]:
             msg = "Base 64 is not supported yet by clarify_path_or_url"
             raise NotImplementedError(msg)
     return file_path, url
+
+
+def resolve_path_or_url_for_reading(path_or_uri: str) -> str:
+    """Resolve a path-or-URI to a single string suitable for file-reading/conversion calls.
+
+    This is a convenience wrapper around `clarify_path_or_url()` that returns a single value:
+    - the http(s) URL if `path_or_uri` is a URL
+    - otherwise a local file path (including normalized `file://...` URIs)
+    """
+    file_path, url = clarify_path_or_url(path_or_uri=path_or_uri)
+    resolved = url or file_path
+    assert resolved is not None
+    return resolved
```

assert in the code ? hmmm, better to raise an error ? 
This would also require a test

---

no: because clarify_path_or_url returns one of them not None (or raises)
the `assert` pleases the type checkers and is optimized out of the code when possible

---

`resolved = url or file_path` is *supposed* to always produce a string here: `clarify_path_or_url()` sets **exactly one** of `url` or `file_path` for every supported input, and raises for unsupported cases (base64).

So `assert resolved is not None` is used to:

- **Enforce an internal invariant**: if both `url` and `file_path` were ever `None` (due to a bug or future change), you fail fast *right here* instead of returning `None` and crashing later in a less obvious place.
- **Help static type checkers** (pyright/mypy): since `url` and `file_path` are typed as `str | None`, `url or file_path` is also inferred as `str | None`. The `assert` narrows it to **`str`** for the return type.

One caveat: Python removes `assert` statements when running with optimizations (`python -O`), so this is primarily a **developer-time safety + type-narrowing** guard, not a user-facing runtime validation.